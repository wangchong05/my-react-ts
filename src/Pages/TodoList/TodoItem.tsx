import React from 'react';
import styled, { StyledProps } from "styled-components";

export interface ItemProps extends StyledProps<any> {}

const TodoItem = (props: ItemProps) => {
  const {item, index, deleteItem, updateItem}: ItemProps = props;
  const { complete, action }: ItemProps = item;

  const styleFun = (index: number) => {
      return index%2 === 0 ? '#f0efef' : 'white';
  }

  const ItemWrapper = styled.div`
    width: 100%;
    display: flex;
    background: ${styleFun(index)};
    text-align: left;
    &:hover {
      background: #e8e6e6;
    }
  `;

  return (
    <ItemWrapper>
    {/* <div> */}
      <div style={{flex: 2}}>{action}</div>
      <div style={{flex: 2}}><span style={{color: complete?'green':'red'}} onClick={()=>{updateItem(index)}}>{complete?'Yes':'No'}</span></div>
      <div style={{flex: 1}}><span onClick={()=>{deleteItem(index)}}>delete</span></div>
    {/* </div> */}
    </ItemWrapper>
  )
}

export default TodoItem;
