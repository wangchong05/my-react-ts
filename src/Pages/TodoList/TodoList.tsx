import React, { useState, useEffect, useRef } from 'react';
import styled, { StyledProps } from "styled-components";
// import TodoItem from './TodoItem';

export interface TodoItemProps extends StyledProps<any> {
  item?: object;
  index?: number;
  deleteItem?: void;
  updateItem?: void;
  complete?: boolean;
}

const listOrigin = [
  {
    action: 'drink',
    complete: false
  }, {
    action: 'eat',
    complete: true
  }
]

const TodoList = () => {

  const [list, setList] = useState(
    listOrigin
  )

  const [updateTime, setUpdateTime] = useState('')

  const myRef = useRef<any>('')

  useEffect(()=>{
    setUpdateTime(new Date().toLocaleString())
  }, [list])

  // 添加
  const addItem = () => {
    if(!myRef.current.value || list.find((e) => { return e.action === myRef.current.value })) {
      // alert('请输入内容或内容重复')
      return
    }
    const newList = [...list, {action: myRef.current.value, complete: false}]
    setList(newList)
    myRef.current.value = ''
  }

  // 删除
  const deleteItem = (index: number) => {
    const newList = [...list]
    newList.splice(index, 1)
    setList(newList)
  }

  // 更新单项
  const updateItem = (index: number) => {
    const newList = [...list]
    newList[index]['complete'] = !newList[index]['complete']
    setList(newList)
  }

  return (
    <div data-testid='todolist'>
      <h3>List</h3>
      <div>Update Time: {updateTime}</div>
      <br/>
      <div style={{display: 'flex', justifyContent: 'space-between'}}>
        <input type="text" ref={myRef} placeholder={'please input'} data-testid='input'/>
        <button onClick={addItem} data-testid='Add'>add</button>
      </div>
      <br/>
      <div>
        {
          list.map((item, index) => {
            return <TodoItem item={item} key={index} index={index} deleteItem={deleteItem} updateItem={updateItem}/>
          })
        }
      </div>

    </div>
  );
}

export interface ItemProps extends StyledProps<any> {}

const TodoItem = (props: ItemProps) => {
  const {item, index, deleteItem, updateItem}: ItemProps = props;
  const { complete, action }: ItemProps = item;

  const styleFun = (index: number) => {
      return index%2 === 0 ? '#f0efef' : 'white';
  }

  const ItemWrapper = styled.div`
    width: 100%;
    display: flex;
    background: ${styleFun(index)};
    text-align: left;
    &:hover {
      background: #e8e6e6;
    }
  `;

  return (
    <ItemWrapper data-testid={"list" + index}>
    {/* <div> */}
      <div style={{flex: 2}}>{action}</div>
      <div style={{flex: 2}}><span data-testid={"complete" + index} style={{color: complete?'green':'red'}} onClick={()=>{updateItem(index)}}>{complete?'Yes':'No'}</span></div>
      <div style={{flex: 1}}><span data-testid={"delete" + index} onClick={()=>{deleteItem(index)}}>delete</span></div>
    {/* </div> */}
    </ItemWrapper>
  )
}

export default TodoList;
