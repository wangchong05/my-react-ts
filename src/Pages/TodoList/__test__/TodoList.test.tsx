import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TodoList from '../TodoList';
// import TodoItem from '../TodoItem';
import renderer from "react-test-renderer";

beforeEach(() => {
  render(<TodoList />);
})
 
describe('TodoList render', () => {// TodoList渲染---------------------(1)
  test('renders visible static elements in the TodoList', () => {
    screen.getByText('List');
    screen.getByText(/Update Time:/);
    expect(screen.getByRole('heading')).toBeInTheDocument();
    expect(screen.getByRole('textbox')).toBeInTheDocument();
    expect(screen.getByRole('button')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('please input')).toBeInTheDocument();

    // expect(screen.getByTestId('todolist')).toMatchSnapshot();// TodoList生成快照
  });
});
 
describe('Input cases', () => {// input输入内容-----------(2)
  test('input text', () => {
    userEvent.type(screen.getByRole('textbox'), 'input text');
    expect(screen.getByDisplayValue('input text')).toBeInTheDocument();
  });
});
 
describe('测列表complete的初始值', () => {// input输入内容-----------(3)
  test('input text', () => {
    const complete0 = screen.getByTestId('complete0');
    expect(complete0).toHaveTextContent("No");
    const complete1 = screen.getByTestId('complete1');
    expect(complete1).toHaveTextContent("Yes");
  });
});


describe('Click cases', () => {// 点击情况-----------(3)
  test('add button is enabled', async () => {
    await userEvent.click(screen.getByTestId('Add'));
    expect(screen.getByTestId('Add')).toBeEnabled();
  });
});


// describe('hover', () => {// ---------(4)
//   test('测试hover时的反应', async () => {
//     const list0 = screen.getByTestId('list0');
//     await userEvent.hover(list0);
//     expect(list0).toHaveStyle({'background': '#e8e6e6'})
//   })
//   test('测试hover时的反应2', async () => {
//     const list1 = screen.getByTestId('list1');
//     await userEvent.hover(list1);
//     expect(list1).toHaveStyle({'background': '#e8e6e6'})
//   })
// })
 
// describe('点击complete的时候,yes和no的切换', () => {// 点击情况-----------(5)
//   test('点击complete的时候,yes和no的切换', async () => {
//     const complete0 = screen.getByTestId('complete0');
//     // expect(complete0).toHaveStyle({'color': 'red'})
//     await userEvent.click(complete0);
//     // expect(complete0).toHaveStyle({'color': 'green'})
//     expect(complete0).toHaveTextContent("Yes");
//   });
// });
 
// describe('点击complete的时候,yes和no的切换2', () => {// 点击情况-----------(6)
//   test('点击complete的时候,yes和no的切换', async () => {
//     const complete1 = screen.getByTestId('complete1');
//     // expect(complete1).toHaveStyle({'color': 'green'})

//     await userEvent.click(complete1);
//     // expect(complete1).toHaveStyle({'color': 'red'})
//     expect(complete1).toHaveTextContent("No");
//   });
// });
 
// describe('点击complete的时候,yes和no的切换2222', () => {// 点击情况-----------(6)
//   test('点击complete的时候,yes和no的切换222', async () => {
//     const tree = renderer.create(<TodoList />).toJSON();
//     console.log('--------------------------tree--------------------------')
//     console.log(tree)
//     console.log(tree.children[5].children)
//     console.log('--------------------------tree--------------------------')
//   });
// });
 
// describe('输入内容点击add的时候,列表长度加一', () => {// 点击情况-----------(7)
//   test('输入内容点击add的时候,列表长度加一', async () => {
//     const tree = renderer.create(<TodoList />).toJSON();
//     const listLength = tree.children[5].children.length;
//     console.log('--------------------------tree1--------------------------')
//     console.log(listLength)
//     userEvent.type(screen.getByRole('textbox'), 'input text');
//     expect(screen.getByDisplayValue('input text')).toBeInTheDocument();

//     await userEvent.click(screen.getByTestId('Add'));
//     const newListLength = tree.children[5].children.length;
//     console.log(newListLength)
//     console.log('--------------------------tree2--------------------------')

//     // expect(screen.getByTestId('input')).toHaveAttribute('type', 'text');
//     // const tree = renderer.create(<TodoList />).toJSON();
//     // console.log('--------------------------tree--------------------------')
//     // console.log(tree)
//     // console.log(tree.children[5].children)
//     // console.log('--------------------------tree--------------------------')
//   });
// });