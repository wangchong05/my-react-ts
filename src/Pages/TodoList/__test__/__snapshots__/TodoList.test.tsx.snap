// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add and delete each item in the list A data was successfully added, and the length of the list was added by one, time changes. 1`] = `
<div
  data-testid="todolist"
>
  <h3>
    List
  </h3>
  <div
    data-testid="update-time"
  >
    Update Time: 
    2021/7/19下午11:14:19
  </div>
  <br />
  <div
    style="display: flex; justify-content: space-between;"
  >
    <input
      data-testid="input"
      placeholder="please input"
      type="text"
    />
    <button
      data-testid="Add"
      disabled=""
    >
      add
    </button>
  </div>
  <br />
  <div
    data-testid="list"
  >
    <div
      class="sc-iTVJFM iFwnAG"
      data-testid="list0"
    >
      <div
        style="flex: 2;"
      >
        drink
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete0"
          style="color: green;"
        >
          Yes
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete0"
        >
          delete
        </span>
      </div>
    </div>
    <div
      class="sc-iBzEeX ddLjoH"
      data-testid="list1"
    >
      <div
        style="flex: 2;"
      >
        eat
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete1"
          style="color: green;"
        >
          Yes
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete1"
        >
          delete
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Add and delete each item in the list Delete each time, the length of the list is reduced by one, time changes. 1`] = `
<div
  data-testid="todolist"
>
  <h3>
    List
  </h3>
  <div
    data-testid="update-time"
  >
    Update Time: 
    2021/7/19下午11:14:20
  </div>
  <br />
  <div
    style="display: flex; justify-content: space-between;"
  >
    <input
      data-testid="input"
      placeholder="please input"
      type="text"
    />
    <button
      data-testid="Add"
      disabled=""
    >
      add
    </button>
  </div>
  <br />
  <div
    data-testid="list"
  >
    <div
      class="sc-gSYDnn eWNiXN"
      data-testid="list0"
    >
      <div
        style="flex: 2;"
      >
        drink
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete0"
          style="color: green;"
        >
          Yes
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete0"
        >
          delete
        </span>
      </div>
    </div>
    <div
      class="sc-laZMeE kmtwSQ"
      data-testid="list1"
    >
      <div
        style="flex: 2;"
      >
        eat
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete1"
          style="color: green;"
        >
          Yes
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete1"
        >
          delete
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Click the status button Click the status button, text and style should be changed, time changes 1`] = `
<div
  data-testid="todolist"
>
  <h3>
    List
  </h3>
  <div
    data-testid="update-time"
  >
    Update Time: 
    2021/7/19下午11:14:18
  </div>
  <br />
  <div
    style="display: flex; justify-content: space-between;"
  >
    <input
      data-testid="input"
      placeholder="please input"
      type="text"
    />
    <button
      data-testid="Add"
      disabled=""
    >
      add
    </button>
  </div>
  <br />
  <div
    data-testid="list"
  >
    <div
      class="sc-iklJeh dBUFYd"
      data-testid="list0"
    >
      <div
        style="flex: 2;"
      >
        drink
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete0"
          style="color: red;"
        >
          No
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete0"
        >
          delete
        </span>
      </div>
    </div>
    <div
      class="sc-jJMGnK hLytLB"
      data-testid="list1"
    >
      <div
        style="flex: 2;"
      >
        eat
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete1"
          style="color: green;"
        >
          Yes
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete1"
        >
          delete
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`add button click cases click the add button directly 1`] = `
<div
  data-testid="todolist"
>
  <h3>
    List
  </h3>
  <div
    data-testid="update-time"
  >
    Update Time: 
    2021/7/19下午11:14:17
  </div>
  <br />
  <div
    style="display: flex; justify-content: space-between;"
  >
    <input
      data-testid="input"
      placeholder="please input"
      type="text"
    />
    <button
      data-testid="Add"
      disabled=""
    >
      add
    </button>
  </div>
  <br />
  <div
    data-testid="list"
  >
    <div
      class="sc-ksluID JMpHA"
      data-testid="list0"
    >
      <div
        style="flex: 2;"
      >
        drink
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete0"
          style="color: red;"
        >
          No
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete0"
        >
          delete
        </span>
      </div>
    </div>
    <div
      class="sc-hBMUJo kWvRdl"
      data-testid="list1"
    >
      <div
        style="flex: 2;"
      >
        eat
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete1"
          style="color: green;"
        >
          Yes
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete1"
        >
          delete
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`add button click cases enter the value is not empty, and then click the add button, time changes 1`] = `
<div
  data-testid="todolist"
>
  <h3>
    List
  </h3>
  <div
    data-testid="update-time"
  >
    Update Time: 
    2021/7/19下午11:14:17
  </div>
  <br />
  <div
    style="display: flex; justify-content: space-between;"
  >
    <input
      data-testid="input"
      placeholder="please input"
      type="text"
    />
    <button
      data-testid="Add"
      disabled=""
    >
      add
    </button>
  </div>
  <br />
  <div
    data-testid="list"
  >
    <div
      class="sc-dlMDgC crzXCx"
      data-testid="list0"
    >
      <div
        style="flex: 2;"
      >
        drink
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete0"
          style="color: red;"
        >
          No
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete0"
        >
          delete
        </span>
      </div>
    </div>
    <div
      class="sc-kfYoZR nWiQl"
      data-testid="list1"
    >
      <div
        style="flex: 2;"
      >
        eat
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete1"
          style="color: green;"
        >
          Yes
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete1"
        >
          delete
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`add button click cases enter the value is not empty, and then click the add button, time changes 2`] = `
<div
  data-testid="todolist"
>
  <h3>
    List
  </h3>
  <div
    data-testid="update-time"
  >
    Update Time: 
    2021/7/19下午11:14:18
  </div>
  <br />
  <div
    style="display: flex; justify-content: space-between;"
  >
    <input
      data-testid="input"
      placeholder="please input"
      type="text"
    />
    <button
      data-testid="Add"
      disabled=""
    >
      add
    </button>
  </div>
  <br />
  <div
    data-testid="list"
  >
    <div
      class="sc-iJCRrE bZYflR"
      data-testid="list0"
    >
      <div
        style="flex: 2;"
      >
        drink
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete0"
          style="color: red;"
        >
          No
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete0"
        >
          delete
        </span>
      </div>
    </div>
    <div
      class="sc-giAqHp ejyxuM"
      data-testid="list1"
    >
      <div
        style="flex: 2;"
      >
        eat
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete1"
          style="color: green;"
        >
          Yes
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete1"
        >
          delete
        </span>
      </div>
    </div>
    <div
      class="sc-ezzafa ldLYJv"
      data-testid="list2"
    >
      <div
        style="flex: 2;"
      >
        input text
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete2"
          style="color: red;"
        >
          No
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete2"
        >
          delete
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`add button click cases enter the value to empty, and then click the add button 1`] = `
<div
  data-testid="todolist"
>
  <h3>
    List
  </h3>
  <div
    data-testid="update-time"
  >
    Update Time: 
    2021/7/19下午11:14:17
  </div>
  <br />
  <div
    style="display: flex; justify-content: space-between;"
  >
    <input
      data-testid="input"
      placeholder="please input"
      type="text"
    />
    <button
      data-testid="Add"
      disabled=""
    >
      add
    </button>
  </div>
  <br />
  <div
    data-testid="list"
  >
    <div
      class="sc-bkbkJK ckstoR"
      data-testid="list0"
    >
      <div
        style="flex: 2;"
      >
        drink
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete0"
          style="color: red;"
        >
          No
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete0"
        >
          delete
        </span>
      </div>
    </div>
    <div
      class="sc-iemWCZ cpULLB"
      data-testid="list1"
    >
      <div
        style="flex: 2;"
      >
        eat
      </div>
      <div
        style="flex: 2;"
      >
        <span
          data-testid="complete1"
          style="color: green;"
        >
          Yes
        </span>
      </div>
      <div
        style="flex: 1;"
      >
        <span
          data-testid="delete1"
        >
          delete
        </span>
      </div>
    </div>
  </div>
</div>
`;
